### PR TITLE
Some fixes

### PR DESCRIPTION
**Pull request by ArminJo**
Hi,
Thanks for the library I plan to use for my project!

I removed trailing blanks, fixed some typos, adjusted some types and documentation and fixed some errors in examples.
I also included a simple CI with Github Actions and enabled the examples for compiling on an ESP32.

May I also suggest to enable discussions in Settings menu and transfer all open issues to discussions.
And you can disable/uncheck unused packages and environment here:
2022-10-12 22_00_05-ArminJo_Arduino-BlueDisplay_ Arduino library for the BlueDisplay App  The App co

### CODE REVIEW

**Comment by WifWaf** [comment]
Hey, 

I'm honestly not familiar with .yml configure and how they work during compilation, so will have a quick read-up in a few days and look over the changes (unwell with covid at the moment, so not much use).

Cheers for taking the time to put this together - it should be fine at a quick glance.

**Comment by ArminJo** [comment]
Results of the CI can be seen here https://github.com/ArminJo/MH-Z19/actions

**Comment by ArminJo** [comment]
Everything OK?

File: examples/Recovery/Recovery.ino [hunk 1]
```diff
@@ -1,37 +1,40 @@
...
+HardwareSerial mySerial(2);
+#else
+#include <SoftwareSerial.h>                               //  Remove if using HardwareSerial or non-uno compatible device
+SoftwareSerial mySerial(RX_PIN, TX_PIN);                   // (Uno example) create device to MH-Z19 serial
+#endif
 
 unsigned long getDataTimer = 0;
 
-void setRange(int range);                                  // Declerations for non-IDE platform
+void setRange(int range);                                  // Decelerations for non-IDE platform
```
Comment (WifWaf) [inline examples/Recovery/Recovery.ino:29]: I believe this should be "declaration".

File: src/MHZ19.cpp [hunk 2]
```diff
@@ -52,13 +54,13 @@ void MHZ19::begin(Stream &serial)
 
 void MHZ19::setRange(int range)
 {
-    if (range > 65000)
+    if (range > __INT_MAX__)
```
Comment (WifWaf) [inline src/MHZ19.cpp:57]: This value is to do with the sensor and I'm not sure __INT_MAX__ will work as expected. Best to leave this as 65000.

File: examples/BasicUsage/BasicUsage.ino [hunk 3]
```diff
@@ -1,22 +1,26 @@
...
-#include "MHZ19.h"                                        
-#include <SoftwareSerial.h>                                // Remove if using HardwareSerial
+#include "MHZ19.h"
 
 #define RX_PIN 10                                          // Rx pin which the MHZ19 Tx pin is attached to
 #define TX_PIN 11                                          // Tx pin which the MHZ19 Rx pin is attached to
 #define BAUDRATE 9600                                      // Device to MH-Z19 Serial baudrate (should not be changed)
 
 MHZ19 myMHZ19;                                             // Constructor for library
+#if defined(ESP32)
```
Comment (WifWaf) [inline examples/BasicUsage/BasicUsage.ino:9]: See the reply to the pull request.

**Comment by WifWaf** [comment]
Yes, sorry for not getting back to you on this.

I had a read over the changes and the corrections are much appreciated. Spelling is not my forte... I've added a few comments which need to be looked at - I assume you can see these? (I'm not experienced with the pull request system).

In regards to the basic example changes, generally, I try to keep the basic example as simple as possible to avoid any confusion as Arduino is a common entry point to programming. Preprocessor directives are a bit more advanced, so for the ESP32 setup, it might be best to comment out the relevant code with a short description (e.g. "for ESP32 users"). 

Let me know your thoughts.

*As a side note, I'm not entirely sure how the GitHub actions/workflow works - is it necessary to include the .yml file?

Cheers,

File: src/MHZ19.cpp (hunk 2 above)
Comment (ArminJo) [reply src/MHZ19.cpp:57]: But you know, that an 16 bit int cannot be greater than 2^15 -1 (32767)?

File: examples/Recovery/Recovery.ino (hunk 1 above)
Comment (ArminJo) [reply examples/Recovery/Recovery.ino:29]: Oh sorry.

**Comment by ArminJo** [comment]
Hi,

the preprocessor directives enables the ESP32 user to compile the examples without errors on the first run (and often, but not in this case) to pass the Github actions. Without it, the user first has to change comments before succesfull running, which is not user friendly in my opinion.
But YES you are right it increases the complexity of the examples!  But I guess the novice user just overreads this preprocessor directives (and other things :-()  and compile it and then modify it. In my ecxperience there are only a few, who try to understand all parts of the code. 
But additional comments are a good idea 😀 .
For my libraries, I decided to have at least one very simple example, and others are more sophisticated, to show users how to program [a bit more professional](https://github.com/Arduino-IRremote/Arduino-IRremote/blob/d51b540cb2ddf1424888d2d9e6b62fe1ef46859d/examples/SimpleReceiver/SimpleReceiver.ino#L68).

And yes you have to include the *.yml file to activate the github actions.

File: src/MHZ19.cpp (hunk 2 above)
Comment (ArminJo) [reply src/MHZ19.cpp:57]: In the manual I found:
`Note: Detection range is 2000 or 5000ppm`
so the check is a bit random. Maybe 
`if (range > 500) { // safety net`
makes more sense?

Comment (WifWaf) [reply src/MHZ19.cpp:57]: The lower threshold does appear to be around 2,000 ppm and the upper 10,000 ppm.

Perhaps the below would cover all use cases?  

if(range >= 500 && range <= 30000) 

20,000 would work as well.

Comment (WifWaf) [reply src/MHZ19.cpp:57]: Of course, with respect to the error messages  :).

if(range < 500 || range > 30000)

Comment (ArminJo) [reply src/MHZ19.cpp]: I took 20000 and commited, this looks more sensible 😀

**Review by WifWaf — COMMENTED** [review]
The range has been made more explicit when being set using setRange(). Previously it did account for the signed bit and would never have caught an invalid value.

A number of formatting and spelling corrections.

Compatibility has been added for the ESP32 to all examples.

All seems fine at a glance.

**Comment by WifWaf** [comment]
Looks greats. I've merged the changes.

There is a more lightweight version of the library on the [PR_2.0.0 branch](https://github.com/WifWaf/MH-Z19/tree/PR_2.0.0), although it's slightly out of date. I've been planning (*tm) on moving it over gradually to the master branch.

I'll make sure the changes are moved to this branch too.

**Comment by ArminJo** [comment]
Thanks! 👍 
I also [refactored the library](https://github.com/ArminJo/MH-Z19/commit/0d35230d9978748c8b54018ae2d776973d7bfd02) to make the processing more transparent and added some functions like printCommand() and debug output. But I am not finished yet.
Best regards
Armin

**Pull request merged**